pkg/tests/ossm/ior_test.go

```diff
@@ -0,0 +1,322 @@
+
+			t.LogStep("Check whether the IOR creates Routes for hosts specified in the Gateway")
+			createSimpleGateway(t)
+			checkSimpleGateway(t)
+		})
+
+		t.NewSubTest("check routes that are not deleted during v2.3 to v2.4 upgrade").Run(func(t test.TestHelper) {
+			if env.GetSMCPVersion().LessThan(version.SMCP_2_4) {
+				t.Skip("This test only applies for v2.3 to v2.4 upgrade")
+			}
+
+			t.Cleanup(func() {
+				oc.RecreateNamespace(t, meshNamespace)
+				setupDefaultSMCP(t, meshNamespace)
+			})
+
+			t.LogStepf("Delete and recreate namespace %s", meshNamespace)
+			oc.RecreateNamespace(t, meshNamespace)
+
+			t.LogStep("Deploy SMCP v2.3")
+			setupV23SMCP(t, meshNamespace, meshName)
+
+			t.LogStep("Check whether IOR creates Routes for hosts specified in the Gateway")
+			createSimpleGateway(t)
+			checkSimpleGateway(t)
+
+			t.LogStepf("Record the Route before the upgrade")
+			before := getRoutes(t, meshNamespace)
+
+			t.LogStep("Upgrade SMCP to v2.4")
+			updateToV24SMCP(t, meshNamespace, meshName)
+
+			t.LogStep("Check the Routes existed afther the upgrade")
+			checkSimpleGateway(t)
+			after := getRoutes(t, meshNamespace)
+
+			t.LogStep("Check the Routes unchanged afther the upgrade")
+			if before[0].Metadata.ResourceVersion != after[0].Metadata.ResourceVersion {
+				t.Fatal("Expect the route to be unchanged, but it is changed after the upgrade")
+			} else {
+				t.LogSuccess("Got the same resourceVersion before and after the upgrade")
+			}
+		})
+
+		t.NewSubTest("check IOR does not delete routes after deleting Istio pod").Run(func(t test.TestHelper) {
+			total := 3
+			nsNames := []string{}
+			gateways := []string{}
+
+			t.Cleanup(func() {
+				if env.GetSMCPVersion().GreaterThanOrEqual(version.SMCP_2_4) {
+					removeIORCustomSetting(t, meshNamespace, meshName)
+				}
+
+				oc.DeleteNamespace(t, nsNames...)
+				oc.ApplyString(t, meshNamespace, GetSMMRTemplate())
+				oc.WaitSMMRReady(t, meshNamespace)
+			})
+
+			t.LogStep("Ensure the IOR enabled")
+			if env.GetSMCPVersion().GreaterThanOrEqual(version.SMCP_2_4) {
+				enableIOR(t, meshNamespace, meshName)
+			}
+
+			t.LogStepf("Create %d Gateways and they are in their own Namespace", total)
+			for i := 0; i < total; i++ {
+				ns := fmt.Sprintf("ns-%d", i)
+				nsNames = append(nsNames, ns)
+				gateways = append(gateways, generateGateway(fmt.Sprintf("gw-%d", i), ns, fmt.Sprintf("www-%d.test.ocp", i)))
+			}
+			oc.CreateNamespace(t, nsNames...)
+			oc.ApplyString(t, "", gateways...)
+
+			t.LogStepf("Update SMMR to include %d Namespaces", total)
+			oc.ApplyString(t, meshNamespace, fmt.Sprintf(`
+apiVersion: maistra.io/v1
+kind: ServiceMeshMemberRoll
+metadata:
+  name: default
+spec:
+  members:
+  - bookinfo
+  - foo
+  - bar
+  - legacy
+  - %s
+  `, strings.Join(nsNames, "\n  - ")))
+
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				routes := getRoutes(t, meshNamespace)
+				if len(routes) == total {
+					t.LogSuccessf("Found all %d Routes", total)
```

Comment:
Why check just the number of routes instead of checking the actual routes? This makes the test fragile. It could fail eventually, at which point, we'll be wasting time trying to figure out what's wrong.